### DISCUSSION TITLE
Pyinstaller to exe

### DISCUSSION THREAD

**Discussion by tanyibo-cnb** [DHI/mikeio#397] [Q&A] [answered]
Why can't I use the script written by mikeio after it is encapsulated into exe.

**Comment by jsmariegaard** [comment]
Could you please provide some more details? What exactly are you trying to do? What version of mikeio? What type of error messages do you get? etc...

**Reply by tanyibo-cnb** [reply]
I wrote a program to read dfsu files and converted it into an EXE program with "pyinstaller -f * * *. Py", but when it is opened, the CMD window will blink back。
The error information is as follows：
***********************************************
File "ctypesinit. py", line 373, in init FileNotFoundError: Could not find module_'C: Users TAN_LAI AppData Local Temp_ME173362\ mikecorelbinlwindows leum' (or on e of its dependencies). Try using the full path with constructor syntax.
The above exception was the direct cause of the following exception: Traceback (most recent call_last): File "0729rewrite_type. py", line 1, in <module>
from mikeio import Dfsu File "<frozen importlib._bootstrap>", line 991, in _find_and_load File "<frozen importlib._bootstrap)", line 975, in _find_and_load_unlocked File "<frozen importlib._bootstrap>", line 671, in _load_unlocked File "PyInstaller(loader (pyimod03_importers. py", line 495, in exec_module File "mikeiol init py", line 26, in <module>
File "<frozen importlib. _bootstrap>", line 991, in _find_and_load File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked File "<frozen importlib. _bootstrap)
line 671, in _load_unlocked File "PyInstaller(loader (pyimod03_importers. py", line 495, in exec_module File "mikeioldfs0. py line 8, in <module File "<frozen importlib._bootstrap>", line 991, in _find_and_load File "<frozen importlib._bootstrap>
line 975, in _find_and load_unlocked File "(frozen importlib.
bootstrap>
line 671, in _load_unlocked File "PyInstaller(loader (pyimod03_importers. py", line 495, in exec_module File "mikecore init py", line 29, in <module>
File "mikecore leum. py", line 1368, in Init File "PyInstallerlloader(pyimod04_ctypes. py", line 55, in_init, pyimod04_ctypes. PyInstallerImportError: Failed to load dynlib/dl1 C: Users(TAN_LAI (AppData((Local(Temp (_ME1733621
mikecorel(binlIwindowsI leum'. Most likely this dynlib/dli was not found when the application was frozen.
[10220] Failed to execute script
0729rewrite_type' due to unhandled exception!
***********************************************

**Comment by lezhongwen** [comment]
I think I have the same problem lol. I guess he used PyInstaller to wrap the script using Mikeio and there was a bug. Does the CMD window blink back? @tanyibo-cnb

**Reply by tanyibo-cnb** [reply]
yes！I don't know how to deal with it.

**Comment by ecomodeller** [comment]
Please be aware that MIKE IO is not a pure python library and depends upon some shared libraries (dlls) in the mikecore library. It seems like the problem with PyInstaller is related to finding these dlls .
![image](https://user-images.githubusercontent.com/614215/182299695-b348519a-dde0-4fe9-bf5e-8780377f6813.png)

**Comment by ecomodeller** [comment, marked as answer]
@PierreBrinkSwiegers you managed to create a working exe with PyInstaller by adding binaries somehow, am I right?

**Reply by PierreBrinkSwiegers** [reply]
@ecomodeller, I was successful in compiling a python script with a mikeio library to an executable by using the --onedir argument with pynistaller. A manual step was required to ensure the dll's are available:
Step 1 running pyinstaller from command line: pyinstaller --onedir script_to_compile.py
Step 2: Manually copy mikecore bin folder from the python installation (\Lib\site-packages\mikecore\bin) into the dist folder (dist\script_to_compile\mikecore\bin)

The executable (compiled with --onedir) will only run if the dist folder is available to the .exe file. In my case it would be cleaner to run pyinstaller using --onefile. I tried doing that and also tried to specify the path to the bin folder explicitly, but that was unsuccessful.
